Dynamically load case details when clicking markers

MartinRiese opened the pull request:
## Product Description

Instead of getting the content of the map marker pop ups from a case list column get it from the details.
The change actually supports both ways for the transition. If the address popup is not defined in a case list column the code will attempt to fetch it from the details. That means the app needs to be updated to define the address popup field in the case details instead of the case list. Any subsequent click on the marker will not fetch the details again.

## Technical Summary

https://dimagi-dev.atlassian.net/browse/USH-3195

## Feature Flag

* USH: Allow use of a map in the case list in Web Apps …

## Safety Assurance

### Safety story

Tested locally:

1. Previously configured app with marker pop upstill works
2. Removing popup config from case list does not show any popup
3. Adding new popup config in case details triggers details request and shows popup
4. After changing popup content using clickable icons and reopening the popup will show new content
5. Adding multiple taps in case details still shows popup the same way.

### Automated test coverage

Not sure how to test that automatically since most of it involves a third party library.

### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

MartinRiese committed:
Dynamically load case details when clicking markers

MartinRiese committed:
Fix naming

MartinRiese committed:
Hide address popup fields on details modal

MartinRiese committed:
Change name so variable do not get shadowed

MartinRiese committed:
Load details every time the popup is opened

MartinRiese commented on corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/controller.js:
```diff
@@ -252,10 +252,12 @@ hqDefine("cloudcare/js/formplayer/menus/controller", function () {
                 obj.html = markdown.render(details[i]);
             }
-            detailModel.push(obj);
+            if (obj.style.displayFormat !== constants.FORMAT_ADDRESS_POPUP) {
```
No super sure about this. For the case list we set widthHint to -1 in FP to hide fields. But for details they are -1. Either way to me it seems more a UI responsibility to hide it.

MartinRiese commented:
Any suggestions for automated test are welcome.

Robert-Costello commented on corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/views.js:
```diff
@@ -869,10 +869,11 @@ hqDefine("cloudcare/js/formplayer/menus/views", function () {
-                                            if (popupIndex >= 0) {
-                                                const popupText = markdown.render(attributes.details[popupIndex]);
-                                                const p = L.popup().setContent(popupText);
+                                            const popupIndexOnClick =
```
By shadowing do you mean the same variable names potentially getting used in other places?

MartinRiese replied:
Yes same variable name as used in outer scope.

Robert-Costello commented:
> Any suggestions for automated test are welcome.

It would be a relatively big effort to create some mock mapbox responses similar to the way we mock formplayer responses. It might be something to consider as a separate ticket or series of tickets in the future. https://github.com/dimagi/commcare-hq/blob/3a66161257ea8c7e8e969506a73d9481d8d50abe/corehq/apps/cloudcare/static/cloudcare/js/formplayer/spec/fake_formplayer.js#L2

Robert-Costello reviewed (approved):
Left one question, but this looks good to me. Will this go on staging for testing?

Jtang-1 commented:
I'm guessing the continued support for address popup configuration in the case list will be just temporary? If so, is the transition plan to just ask app builders to move the configuration?

MartinRiese commented:
@Jtang-1 So far it has been possible to use the Address Popup format in both case list and detail. Although, only in case list it caused the popup so far. I will investigate adding validation to not use it on the list.

MartinRiese committed:
Update validation for address popup configuration

* Check that address popup configuration is not used in case list

nospame commented on corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/views.js:
```diff
@@ -863,6 +863,23 @@ hqDefine("cloudcare/js/formplayer/menus/views", function () {
+                                        const caseId = model.get('id');
+                                        const urlObject = formplayerUtils.currentUrlToObject();
+                                        urlObject.addSelection(caseId);
+                                        const fetchingDetails = FormplayerFrontend.getChannel().request("entity:get:details", urlObject, false, false, false, true);
```
Doesn't this function only take 4 parameters?
https://github.com/dimagi/commcare-hq/blob/79f02b5e0ec6be0839a230f547cf6be7334162c3/corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/api.js#L227

MartinRiese replied:
Good catch. I just copied another call I found and modified it. @Robert-Costello was there a reason you used 5 arguments [here](https://github.com/dimagi/commcare-hq/blob/23943f827d095de9a25a02703e3ca668fad8348a/corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/views.js#L333)?

Robert-Costello replied:
It was for a flag that indicated that the request was coming from a clickable icon. It should have been removed when we went a different route with case list filtering. If you want you could remove the extra argument here or I can make a separate PR.

MartinRiese replied:
I removed them

nospame reviewed (commented):
re: @Jtang-1 's point, based on the added validator it seems like the plan _is_ to move to only the new case detail format for address popup, right? Is there a followup planned or will the `address_popup_in_long` validator stay in place indefinitely?

MartinRiese committed:
Refactor "entity:get:details" request parameters in views.js

Removed redundant true parameter in the last position.

MartinRiese committed:
Remove map_popup field from case tile templates

map_popup was a hidden field used to configure a field with the
"Address Popup" format. This should now be done in the case details.

MartinRiese committed:
Add validator tests

* Small refactor for common app and module creation
* Add test for duplicate Address fields
* Add test for Address Popup field in case list

Jtang-1 commented:
It sounds like the plan is to
With this PR, 
- continue to support existing builds where address pop up is defined in case list. 
- don't allow new builds where address popup is defined in case list (Does address popup need to be removed from "format" options on case list UI?)

Followup PR 
- removes the backend code that supports address popups defined in case list
- How do we ensure all builds have been converted to using case detail for address popup?

MartinRiese committed:
Remove details tab models if they do not have visible styles

MartinRiese committed:
Linting fix

MartinRiese committed:
Merge branch 'master' into riese/dynamic_map_pop_up

MartinRiese commented:
@Jtang-1 I think it really is only used in the BHA apps and woody knows about the change. So I am certain he will track it.

Jtang-1 commented on corehq/apps/app_manager/helpers/validators.py:
```diff
@@ -565,10 +565,13 @@ def validate_case_list_field_actions(self):
     __invalid_tile_configuration_type: str = "invalid tile configuration"
-
     __invalid_clickable_icon_configuration: str = "invalid clickable icon configuration"
+    __deprecated_popup_configuration: str = "deprecated popup configuration"
```
Is this used anywhere?

MartinRiese replied:
I meant to. Good catch

Jtang-1 commented on corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/controller.js:
```diff
@@ -179,7 +179,12 @@ hqDefine("cloudcare/js/formplayer/menus/controller", function () {
-        var detailObjects = model.models;
+        var detailObjects = model.filter(function(d){
+            const styles = d.get('styles');
+            const visibleStyle = _.find(styles, s => s.displayFormat !== constants.FORMAT_ADDRESS_POPUP);
```
could `styles` potentially be undefined?

Jtang-1 commented:
Also noting - the validation will make it such that apps currently using address-popup on case list will not be able to make a new build without moving it to case details. I'm assuming that's been communicated to delivery

MartinRiese committed:
Add migration command

* Add command to migrate address popup fields from details short (case list) to details long (case details)

MartinRiese committed:
Temporarily disable address popup validation

MartinRiese commented on corehq/apps/app_manager/helpers/validators.py:
```diff
@@ -585,6 +588,20 @@ def _validate_fields_with_format(
+        errors: list
+    ):
+        fields_with_address_format = \
+            [c.field for c in self.module.case_details.short.columns if c.format == self.__address_popup]
```
yes. changed it.

MartinRiese commented:
@Jtang-1 @nospame Change of plans. I disabled the validator for now and added a migration command. Since the UI code supports address popup in both the case list and case detail, we can do the deployment in the following steps.

1. Deploy
2. Run the command
3. Wait a week. Hopefully everybody that needs to creates a new version or their app
4. Next deploy
  * Add validator back in
  * Possibly remove the code that support address popups in case lists
  * Add an actual migration that runs the command but will skip all already processed projects

MartinRiese committed:
Use set instead of list

* A set is sufficient here since we only care about the presents of an address popup, not how many

MartinRiese committed:
Disable test

* Since the validator has been temporarily disabled, also disable the test for it.

MartinRiese committed:
Add test for migrate_address_popup command

Jtang-1 reviewed (approved):
I see there's a bunch of code commented out to be reapplied after the migration. My preference is would be to just delete the code and put it in another branch. But other than that, looks good.
